### PR TITLE
Improve Snowflake syntax support

### PR DESCRIPTION
**Pull request by kawashiro**
### Brief summary of the change made

* Extend Snowflake syntax with FOR-IN loop support
* Extend snowflake syntax with bind variables support in IDENTIFIER() statement

### Are there any other side effects of this change that we should be aware of?
Not expected

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 1]
```diff
@@ -7827,3 +7829,38 @@ class AlterMaskingPolicySegment(BaseSegment):
...
+
+class ForInLoopSegment(BaseSegment):
+    """FOR...IN...DO...END FOR statement.
+
+    https://docs.snowflake.com/en/developer-guide/snowflake-scripting/loops#for-loop
+    """
+
+    type = "for_in_statement"
+
+    match_grammar = OneOf(
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_snowflake.py:7842]: Should this not be a Sequence at the top level?

Comment (kawashiro) [reply src/sqlfluff/dialects/dialect_snowflake.py:7842]: It looks like there is a mess with terminators in Delimited() statements and FOR statement itself. Unfortunately, could not find more pretty workaround of that.

Comment (alanmcruickshank) [reply src/sqlfluff/dialects/dialect_snowflake.py:7842]: I think it *should* be possible, but I'll have a go and see if I can make it work that way.

Comment (alanmcruickshank) [reply src/sqlfluff/dialects/dialect_snowflake.py:7842]: Yep - there's something strange going on here. I think it's actually a problem with `ScriptingBlockStatementSegment`. I'm going to have a go at fixing it.

Comment (alanmcruickshank) [reply src/sqlfluff/dialects/dialect_snowflake.py:7842]: Yeah I think there was a bug in the terminator support. I think I've found an fixed it - it also threw up that there's a problem with one of our existing test cases, which I've also solved. @WittierDinosaur I'd love a review given I've added quite a lot of code to this PR.

File: test/fixtures/dialects/snowflake/begin_end.sql [hunk 2]
```diff
@@ -1,7 +1,11 @@
-begin;
+-- NOTE: This is a loop BEGIN, and not a transaction BEGIN,
+-- because BEING is  NOT followed immediately by a ";"
```
Comment (WittierDinosaur) [inline test/fixtures/dialects/snowflake/begin_end.sql:2]:
```suggestion
-- because BEGIN is  NOT followed immediately by a ";"
```

**Review by alanmcruickshank — APPROVED** [review]

**Pull request merged**